maint/local_python/binding_c.py

```diff
@@ -397,70 +638,212 @@ def dump_function(func, map_name):
+        G.out.append("MPID_THREAD_CS_EXIT(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);")
     G.out.append("return mpi_errno;")
     G.out.append("")
     G.out.append("fn_fail:")
     dump_mpi_fn_fail(func, mapping)
     G.out.append("goto fn_exit;")
-    G.out.append("DEDENT")
+
+def push_impl_decl(func, impl_name=None):
+    if not impl_name:
+        impl_name = re.sub(r'^MPIX?_', 'MPIR_', func['name']) + "_impl"
+    if func['impl_param_list']:
+        params = ', '.join(func['impl_param_list'])
+        if func['dir'] == 'coll' and not RE.match(r'MPI_I', func['name']):
+            params = params + ", MPIR_Errflag_t *errflag"
+    else:
+        params="void"
+    G.impl_declares.append("int %s(%s);" % (impl_name, params))
+
+def dump_CHECKENUM(var, errname, t, type="ENUM"):
+    val_list = t.split()
+    if type == "ENUM":
+        cond_list = []
+        for val in val_list:
+            cond_list.append("%s != %s" % (var, val))
+        cond = ' && '.join(cond_list)
+    elif type == "MASK":
+        cond = "%s != (%s & (%s))" % (var, var, ' | '.join(val_list))
+    dump_if_open(cond)
+    G.out.append("MPIR_ERR_SET1(mpi_errno, MPI_ERR_ARG, \"**%s\", \"**%s %%d\", %s);" % (errname, errname, var))
+    G.out.append("goto fn_fail;")
+    dump_if_close()
+
+def dump_body_coll(func):
+    # collectives call either MPID or MPIR depending on CVARs
+    RE.match(r'MPI_(\w+)', func['name'])
+    name = RE.m.group(1)
+
+    if name.startswith('I'):
+        G.out.append("MPIR_Request *request_ptr = NULL;")
+
+    cond_a = "MPIR_CVAR_DEVICE_COLLECTIVES == MPIR_CVAR_DEVICE_COLLECTIVES_all"
+    cond_b1 = "MPIR_CVAR_DEVICE_COLLECTIVES == MPIR_CVAR_DEVICE_COLLECTIVES_percoll"
+    cond_b2 = "MPIR_CVAR_%s_DEVICE_COLLECTIVE" % name.upper()
+
+    args = ", ".join(func['impl_arg_list'])
+    if not name.startswith('I'):
+        G.out.append("MPIR_Errflag_t errflag = MPIR_ERR_NONE;")
+        args = args + ", " + "&errflag"
+    G.out.append("if (%s || (%s && %s)) {" % (cond_a, cond_b1, cond_b2))
+    G.out.append("    mpi_errno = MPID_%s(%s);" % (name, args))
+    G.out.append("} else {")
+    G.out.append("    mpi_errno = MPIR_%s_impl(%s);" % (name, args))
     G.out.append("}")
+    dump_error_check("")
+    if name.startswith('I'):
+        G.out.append("if (!request_ptr) {")
+        G.out.append("    request_ptr = MPIR_Request_create_complete(MPIR_REQUEST_KIND__COLL);")
+        G.out.append("}")
+        G.out.append("*request = request_ptr->handle;")
+
+def dump_body_impl(func, prefix='mpir'):
+    # mpi_errno = MPIR_Xxx_impl(...);
+    if '_has_handle_out' in func:
+        for p in func['_has_handle_out']:
+            (name, kind) = (p['name'], p['kind'])
+            mpir_type = G.handle_mpir_types[kind]
+            G.out.append("%s *%s_ptr = NULL;" % (mpir_type, name))
+            if kind in G.handle_NULLs:
+                G.out.append("*%s = %s;" % (name, G.handle_NULLs[kind]))
+    elif RE.match(r'mpi_type_', func['name'], re.I):
```

Comment:
We should use the long form `re.IGNORECASE` to make intent clearer.

Reply:
Okay

Reply:
mark as fixed (once pushed)